core/ir/aarch64/codec.c

```diff
@@ -1556,6 +1557,113 @@ encode_opnd_cond(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out)
     return encode_opnd_int(12, 4, false, 0, 0, opnd, enc_out);
 }
 
+/* fpimm8: immediate operand for SIMD fmov */
+
+static inline bool
+decode_opnd_fpimm8(uint enc, int opcode, byte *pc, OUT opnd_t *opnd)
+{
+    /* See Arm Architecture Reference Manual
+     *
+     * Immediate is encoded as 8 bits. Bits 5->9 and 16->18. LSB is bit 5:
+     * imm8 = a:b:c:d:e:f:g:h (LSB)
+     *
+     * Half-precision (v8.2)
+     * --------------
+     *
+     * imm16 = imm8<7>:NOT(imm8<6>):Replicate(imm8<6>,2):imm8<5:0>:Zeros(6);
+     *         a:~b:bb:cdefgh:000000
+     *
+     * datasize = if Q == '1' then 128 else 64;
+     * imm = Replicate(imm16, datasize DIV 16);
+     *     = imm16:imm16:imm16:imm16                         (Q=0 -> 64)
+     *     = imm16:imm16:imm16:imm16:imm16:imm16:imm16:imm16 (Q=1 -> 128)
+     *
+     * Single-precision (TODO)
+     * ----------------
+     * Assume cmode = 1111 and op = 0
+     *
+     * imm32 = imm8<7>:NOT(imm8<6>):Replicate(imm8<6>,5):imm8<5:0>:Zeros(19);
+     *         a:~b:bbbbb:cdefgh:0000000000000000000
+     *
+     * imm64 = Replicate(imm32, 2);
+     *       = a:~b:bbbbb:cdefgh:0000000000000000000 a:~b:bbbbb:cdefgh:0000000000000000000
+     *
+     * datasize = if Q == '1' then 128 else 64;
+     * imm = Replicate(imm64, datasize DIV 64);
+     *     = imm64       (Q=0)
+     *     = imm64:imm64 (Q=1)
+     */
+    union {
+        __fp16 f;
```

Comment:
As discussed, this needs some way to construct a type known to the x86 compiler for a64-on-x86.  Not sure what the best solution is: build a 32-bit float from the pieces manually; find some half-float support in gcc and clang; sthg.

Reply:
After looking at half-precision support on GCC and Clang/LLVM for X86 and considering various options including intrinsics for conversion, I decided to build 32-bit float pieces manually. The variation in h/w support and relatively immature compiler support suggests that doing this manually for now is the best low-maintenance approach.